File: scm-csharp/scheme/primitives/PrimitiveOpenOutputZipBytevector.cs
using System.IO.Compression;

namespace scheme;

public class PrimitiveOpenOutputZipBytevector : Primitive
{
    public override string Name() => "open-output-zip-bytevector";

    public override string Info() =>
        "Syntax: (open-output-zip-bytevector)\n" +
        "Library: (scm zip)\n" +
        "Description: Creates a new in-memory ZIP archive and returns a ZIP writer object. " +
        "After writing entries and calling close-output-zip, use get-output-zip-bytevector to retrieve the bytes.\n" +
        "Example:\n" +
        "  (let ((z (open-output-zip-bytevector)))\n" +
        "    (zip-add-text-entry z \"hello.txt\")\n" +
        "    (close-output-zip z)\n" +
        "    (get-output-zip-bytevector z))";

    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 0, 0);
        var mem = new MemoryStream();
        var output = new ZipOutput
        {
            mem_strm = mem,
            zip = new ZipArchive(mem, ZipArchiveMode.Create, leaveOpen: true)
        };
        return new NativeValue(output);
    }
}

File: scm-java/src/main/scheme/primitives/PrimitiveOpenOutputZipBytevector.java
package scheme.primitives;

import java.io.ByteArrayOutputStream;
import java.util.zip.ZipOutputStream;

import scheme.NativeValue;
import scheme.Primitive;
import scheme.SourcePos;
import scheme.ZipOutputHolder;

public class PrimitiveOpenOutputZipBytevector extends Primitive {
    @Override
    public String name() {
        return "open-output-zip-bytevector";
    }

    @Override
    public String info() {
        return "Syntax: (open-output-zip-bytevector)\n" +
               "Library: (scm zip)\n" +
               "Description: Creates a new in-memory ZIP archive and returns a ZIP writer object. " +
               "After writing entries and calling close-output-zip, use get-output-zip-bytevector to retrieve the bytes.\n" +
               "Example:\n" +
               "  (let ((z (open-output-zip-bytevector)))\n" +
               "    (zip-add-text-entry z \"hello.txt\")\n" +
               "    (close-output-zip z)\n" +
               "    (get-output-zip-bytevector z))";
    }

    @Override
    public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 0, 0);
        var mem = new ByteArrayOutputStream();
        var zip = new ZipOutputStream(mem);
        return new NativeValue(new ZipOutputHolder(zip, mem));
    }
}
